File: app/src/main/java/com/example/win/easy/repository/db/dao/SongXSongListDao.java
package com.example.win.easy.repository.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.ForeignKey;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.RoomWarnings;

import com.example.win.easy.repository.db.data_object.SongListDO;
import com.example.win.easy.repository.db.data_object.SongDO;
import com.example.win.easy.repository.db.data_object.SongXSongListDO;

import java.util.Collection;
import java.util.List;

@Dao
public interface SongXSongListDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(SongXSongListDO songXSongListDO);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Collection<SongXSongListDO> songXSongListDOs);

    //以下为Repo涉及方法
    @Query("SELECT * FROM SongXSongListDO")
    List<SongXSongListDO> getAllRelation();

    @Query("SELECT * FROM SongXSongListDO " +
            "WHERE SongXSongListDO.songId=:songId " +
            "AND SongXSongListDO.songListId=:songListId")
    SongXSongListDO findById(Long songId, Long songListId);

    @Query("SELECT SongXSongListDO.songId, songListId FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id = SongXSongListDO.songId " +
            "WHERE SongDO.remoteId is not null")
    List<SongXSongListDO> findAllDataOnWeb();

    @Query("SELECT SongXSongListDO.songId, songListId FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id = SongXSongListDO.songId " +
            "WHERE SongDO.remoteId is null")
    List<SongXSongListDO> findAllDataOnLocal();

    @Query("SELECT * " +
            "FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id=SongXSongListDO.songId " +
            "WHERE SongXSongListDO.songListId=:songListId")
    @SuppressWarnings(RoomWarnings.CURSOR_MISMATCH)
    List<SongDO> getSongsOf(Long songListId);
    /***********************/


    @Query("SELECT * FROM SongXSongListDO")
    LiveData<List<SongXSongListDO>> findAllSongXSongLists();

    @Query("SELECT * " +
            "FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id=SongXSongListDO.songId " +
            "WHERE SongXSongListDO.songListId=:songListId")
    @SuppressWarnings(RoomWarnings.CURSOR_MISMATCH)
    LiveData<List<SongDO>> findAllSongsForSongListById(Long songListId);

    @Query("SELECT * " +
            "FROM SongListDO INNER JOIN SongXSongListDO " +
            "ON SongListDO.id=SongXSongListDO.songListId " +
            "WHERE SongXSongListDO.songId=:songId")
    @SuppressWarnings(RoomWarnings.CURSOR_MISMATCH)
    List<SongListDO> findAllSongListsForSongById(Long songId);

    @Delete
    void delete(SongXSongListDO songXSongListDO);

}

File: app/src/main/java/com/example/win/easy/repository/db/dao/SongXSongListDao.kt
package com.example.win.easy.repository.db.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.example.win.easy.repository.db.data_object.SongDO
import com.example.win.easy.repository.db.data_object.SongListDO
import com.example.win.easy.repository.db.data_object.SongXSongListDO

@Dao
interface SongXSongListDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(songXSongListDOs: Collection<SongXSongListDO>): List<Long>

    //以下为Repo涉及方法
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(songXSongListDO: SongXSongListDO): Long

    @get:Query("SELECT * FROM SongXSongListDO")
    val allRelation: List<SongXSongListDO?>?

    @Query("SELECT * FROM SongXSongListDO " +
            "WHERE SongXSongListDO.songId=:songId " +
            "AND SongXSongListDO.songListId=:songListId")
    fun findById(songId: Long?, songListId: Long?): SongXSongListDO?

    @Query("SELECT SongXSongListDO.songId, songListId FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id = SongXSongListDO.songId " +
            "WHERE SongDO.remoteId is not null")
    fun findAllDataOnWeb(): List<SongXSongListDO?>?

    @Query("SELECT SongXSongListDO.songId, songListId FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id = SongXSongListDO.songId " +
            "WHERE SongDO.remoteId is null")
    fun findAllDataOnLocal(): List<SongXSongListDO?>?

    @Query("SELECT * " +
            "FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id=SongXSongListDO.songId " +
            "WHERE SongXSongListDO.songListId=:songListId")
    fun getSongsOf(songListId: Long?): List<SongDO?>?

    /** */
    @Query("SELECT * FROM SongXSongListDO")
    fun findAllSongXSongLists(): LiveData<List<SongXSongListDO?>?>?

    @Query("SELECT * " +
            "FROM SongDO INNER JOIN SongXSongListDO " +
            "ON SongDO.id=SongXSongListDO.songId " +
            "WHERE SongXSongListDO.songListId=:songListId")
    fun findAllSongsForSongListById(songListId: Long?): LiveData<List<SongDO?>?>?

    @Query("SELECT * " +
            "FROM SongListDO INNER JOIN SongXSongListDO " +
            "ON SongListDO.id=SongXSongListDO.songListId " +
            "WHERE SongXSongListDO.songId=:songId")
    fun findAllSongListsForSongById(songId: Long?): List<SongListDO?>?

    @Delete
    fun delete(songXSongListDO: SongXSongListDO?)
}